### ISSUE TITLE
Cluster failure after migrating nodepool 'master not discovered or elected yet, an election requires at least 2 nodes with ids'

### ISSUE THREAD

**Issue by wesleyjconnorsesame** [Opster/opensearch-k8s-operator#266]
With a successfully running cluster with 3 master I needed to migrate to a new nodepool (unrelated)
However after migration the cluster was in an unusable state a single master was running `cluster-master-0` with the following error in the logs

[cluster-masters-0] master not discovered or elected yet, an election requires at least 2 nodes with ids from [LNCCMf06T4-3j8DIibFf2g, Y4A36mMzRNeVtsWy7A2kEw, RLoWaFxKSDy04rOanWh50Q],have discovered [{cluster-masters-0}{Y4A36mMzRNeVtsWy7A2kEw}{s43hlqGfRfq1IaTL7kXAtw}{cluster-masters-0} {10.104.8.14:9300}{dm}{shard_indexing_pressure_enabled=true}] which is not a quorum; discovery will continue using [] from hosts providers and [{cluster-masters-0}{Y4A36mMzRNeVtsWy7A2kEw}{s43hlqGfRfq1IaTL7kXAtw}{cluster-masters-0}{10.104.8.14:9300}{dm}{shard_indexing_pressure_enabled=true}]from last-known cluster state; node term 6, last-accepted version 50 in term 5

I think i need to add a  podDisruptionBudget to ensure 2 pods are running while doing a migration of this sort, but I cannot see that option in the cluster config

**Comment by dickescheid** [comment]
I have the same problem while testing:

https://github.com/Opster/opensearch-k8s-operator/issues/261

**Comment by swoehrl-mw** [comment]
Hi @wesleyjconnorsesame . What did you do to migrate to a new nodepool? What should work (without having tested it):
* Add a new nodepool to the cluster
* Wait for it to be integrated
* Reduce replica size of old nodepool to 0 and wait for operator to do the scaling (will take time as it waits for cluster green health between steps)
* Remove old nodepool completely

Did you activate the Smartscaler (`spec.confMgmt.smartScaler`)? If not then removing a nodepool will just delete the pods of that pool without the scaler doing its work.

**Comment by wesleyjconnorsesame** [comment]
> Hi @wesleyjconnorsesame . What did you do to migrate to a new nodepool? What should work (without having tested it):
> 
> * Add a new nodepool to the cluster
> * Wait for it to be integrated
> * Reduce replica size of old nodepool to 0 and wait for operator to do the scaling (will take time as it waits for cluster green health between steps)
> * Remove old nodepool completely
> 
> Did you activate the Smartscaler (`spec.confMgmt.smartScaler`)? If not then removing a nodepool will just delete the pods of that pool without the scaler doing its work.

Thanks, yes basically that was the process

I did not have the smartScaler applied, I see it referenced  [here](https://github.com/Opster/opensearch-k8s-operator/blob/master/opensearch-operator/examples/opensearch-cluster.yaml#L25) but not any docs for it

I have added it to a test cluster now, but dont see any changes, do I need to recreate cluster for it to apply?

**Comment by swoehrl-mw** [comment]
> I did not have the smartScaler applied, I see it referenced [here](https://github.com/Opster/opensearch-k8s-operator/blob/master/opensearch-operator/examples/opensearch-cluster.yaml#L25) but not any docs for it

Good point. We need to improve the docs for that. 

> do I need to recreate cluster for it to apply?

As long as no scaling operation was active while you activated the option, it should work on an existing cluster. You can always check the logs of the operator pod if there are any problems reported. And you can also do a `kubectl describe opensearchcluster ...` to see if the operator reported any events.

> Is the above state at all recoverable?

If the PVCs are still there you could try readding the old nodepool and see if the pods come back up.

**Comment by jinchengsix** [comment]
I got the same error, not sure if it's the the same， what I did is as follow
1. Deploy the cluster normally with examples/opensearch-cluster.yaml
2. All resources are Running fine
3. Delete the statefulset [cluster-masters]
4. Pod cluster-masters-0 is being recreated but not Ready , no matter how long I wait
5. kubectl -n opensearch logs -f  cluster-masters-0 , I can see the error as above

**opensearch-cluster.yaml**
```apiVersion: opensearch.opster.io/v1
kind: OpenSearchCluster
metadata:
  name: cluster
spec:
  general:
    version: 1.3.0
    httpPort: 9200
    vendor: opensearch
    serviceName: cluster
    pluginsList: ["repository-s3"," https://github.com/aiven/prometheus-exporter-plugin-for-opensearch/releases/download/1.3.0.0/prometheus-exporter-1.3.0.0.zip"]
  dashboards:
    version: 1.3.0
    enable: true
    replicas: 1
    resources:
      requests:
         memory: "1Gi"
         cpu: "500m"
      limits:
         memory: "1Gi"
         cpu: "500m"
  confMgmt:
    smartScaler: true
  nodePools:
    - component: masters
      replicas: 3
      diskSize: "50Gi"
      NodeSelector:
      resources:
         requests:
            memory: "1Gi"
            cpu: "500m"
         limits:
            memory: "1Gi"
            cpu: "500m"
      roles:
        - "master"
    - component: nodes
      replicas: 3
      diskSize: "80Gi"
      NodeSelector:
      resources:
         requests:
            memory: "1Gi"
            cpu: "500m"
         limits:
            memory: "1Gi"
            cpu: "500m"
      roles:
        - "data"
    - component: ingest
      replicas: 3
      diskSize: "30Gi"
      NodeSelector:
      resources:
         requests:
            memory: "1Gi"
            cpu: "500m"
         limits:
            memory: "1Gi"
            cpu: "500m"
      roles:
        - "ingest"
```


**screenshot**
![df45133172244acb31fcf92183e8fca](https://user-images.githubusercontent.com/39333013/186877482-18ee8538-7fdc-4124-875c-0c66d1258843.jpg)
![image](https://user-images.githubusercontent.com/39333013/186877622-2bddec76-0d8b-4f4c-b7b4-9ca9fc4f5607.png)

**Comment by swoehrl-mw** [comment]
@wesleyjconnorsesame. In regards to Downsizing not happening with enabled smartscaler: This is likely the same cause as in #227, PR is on the way.

@jinchengsix: I think I understand now what is happening: The statefulsets for the nodepools are configured with a podManagementPolicy of `OrderedReady`. This means kubernetes waits for the first pod to be ready before starting the second. This is normally a good thing as it helps the operator do operations in a rolling manner while always keeping the cluster in a working state. During initial cluster setup the operator works around the quorum problem by launching a separate bootstrap pod to faciliate that. But all this means that for your case the cluster will not come back up as the first pod never gets ready without being able to form a quorum. Not sure if this is a situation we want to handle in the operator. If you feel so please open a new issue so we have it on the roadmap as a potential enhancement.
As a hack to recover a cluster in this situation you can try to manually edit the statefulset (`cluster-masters` in your case) and temporarily set the `podManagementPolicy` to `Parallel`.

**Comment by mahdiG** [comment]
> But all this means that for your case the cluster will not come back up as the first pod never gets ready without being able to form a quorum.

Hi. I've been facing this exact problem. It means if the pods are deleted for any reason, opensearch cluster never comes back up again and it's a big issue. In my experiments, I deleted only one of 3 cluster_manager (master) pods of the statefulset and it deleted other replicas too!! I don't know if it's expected but in that case, I can't have even one node failure or the entire cluster is down forever!
It's impossible to ensure disaster never happens; but it's important that if it does and all pods go down, they can come back up and cluster becomes functional again :)

**Comment by mahdiG** [comment]
> As a hack to recover a cluster in this situation you can try to manually edit the statefulset (`cluster-masters` in your case) and temporarily set the `podManagementPolicy` to `Parallel`.

podManagementPolicy field is immutable statefulset is created.

**Comment by ahmedrshdy** [comment]
I have the same exact issue

How to reproduce:

Use the following yaml to deploy a new cluster

apiVersion: opensearch.opster.io/v1
kind: OpenSearchCluster
metadata:
  name: doha-opsearch
  namespace: open-search

spec:
  general:
    serviceName: doha-dic-poc
    setVMMaxMapCount: true
    version: 2.2.1
    # pluginsList: ["opensearch-knn", "opensearch-ml", "opensearch-notifications", "opensearch-alerting"]
    # pluginsList: ["opensearch-alerting"]
  dashboards:
    enable: true
    version: 2.2.1
    replicas: 1
    resources:
      requests:
         memory: "512Mi"
         cpu: "200m"
      limits:
         memory: "512Mi"
         cpu: "200m"
  nodePools:
    - component: masters
      replicas: 3
      diskSize: "6Gi"
      jvm: -Xmx8g -Xms8g
      NodeSelector:
      resources:
         requests:
            memory: "8Gi"
            cpu: "2"
        #  limits:
        #     memory: "2Gi"
        #     cpu: "500m"
      roles:
        - "data"
        - "cluster_manager"

After successful deployment, manually terminate any of the master nodes.

The terminated node is re-provisioned as expected but stuck in unready state.

Digging into logs of this unready node reveals the following error:

`luster-manager not discovered or elected yet, an election requires a node with id [jP-5_H9nSEyqCcrTEuX3sg]z

**Comment by swoehrl-mw** [comment]
> `podManagementPolicy` field is immutable after statefulset is created.

Right, sorry, didn't think of this. 

@ahmedrshdy 
> The terminated node is re-provisioned as expected but stuck in unready state.

I cannot reproduce this. Deploying a cluster and then deleting a master pod will recreate the pod and after about two minutes the pod was healthy again in my experiments.

@mahdiG 
> Hi. I've been facing this exact problem. It means if the pods are deleted for any reason, opensearch cluster never comes back up again and it's a big issue. In my experiments, I deleted only one of 3 cluster_manager (master) pods of the statefulset and it deleted other replicas too!! I don't know if it's expected but in that case, I can't have even one node failure or the entire cluster is down forever!
It's impossible to ensure disaster never happens; but it's important that if it does and all pods go down, they can come back up and cluster becomes functional again :)

The case with deleting just one pod I could not reproduce. In my experiments the other pods stayed up.
But I agree that if all the pods are deleted at the same time the cluster currently does not come back up again. 

As this issue is already a mix of discussion I've created #289 to track this problem separately.

**Comment by ahmedrshdy** [comment]
> @ahmedrshdy 
> > The terminated node is re-provisioned as expected but stuck in unready state.
> 
> I cannot reproduce this. Deploying a cluster and then deleting a master pod will recreate the pod and after about two minutes the pod was healthy again in my experiments.
> 
@swoehrl-mw 
First thanks for your reply, second can you share the configuration you have used that is able to restore a terminated master pod?

**Comment by swoehrl-mw** [comment]
@ahmedrshdy 

```yaml
apiVersion: opensearch.opster.io/v1
kind: OpenSearchCluster
metadata:
  name: my-cluster
  namespace: default
spec:
  general:
    version: 2.2.1
    httpPort: 9200
    vendor: opensearch
    serviceName: my-cluster
  dashboards:
    version: 2.2.1
    enable: true
    replicas: 1
    resources:
      requests:
         memory: "1Gi"
         cpu: "500m"
      limits:
         memory: "1Gi"
         cpu: "500m"
  nodePools:
    - component: masters
      replicas: 3
      diskSize: "30Gi"
      NodeSelector:
      resources:
         requests:
            memory: "1Gi"
            cpu: "500m"
         limits:
            memory: "2Gi"
            cpu: "500m"
      roles:
        - "cluster_manager"
        - "data"
```
(with Operator installed via helm chart version 2.0.4)
After the cluster was fully formed I ran `kubectl delete pod my-cluster-masters-2`. The pod was recreated and became healthy.
I'm testing this on a local k3d cluster.

**Comment by mleklund** [comment]
> `podManagementPolicy` field is immutable after statefulset is created.

Ran into this issue.  I scaled the operator to zero replicas, then deleted the statefulset with `--cascade=orphan`, then recreated the statefulset with `podManagementPolicy` set to `Parallel`.  then once recovered, delete the statefulset again and switch it back to `OrderedReady`.

**Comment by KannappanSomu** [comment]
> > `podManagementPolicy` field is immutable after statefulset is created.
> 
> Ran into this issue. I scaled the operator to zero replicas, then deleted the statefulset with `--cascade=orphan`, then recreated the statefulset with `podManagementPolicy` set to `Parallel`. then once recovered, delete the statefulset again and switch it back to `OrderedReady`.

I can confirm this hack works. I ran into same issue today and recovered as stated above.

**Comment by idanl21** [comment]
Thanks @KannappanSomu 
closing the issue

**Comment by arve0** [comment]
> > `podManagementPolicy` field is immutable after statefulset is created.
> 
> Ran into this issue. I scaled the operator to zero replicas, then deleted the statefulset with `--cascade=orphan`, then recreated the statefulset with `podManagementPolicy` set to `Parallel`. then once recovered, delete the statefulset again and switch it back to `OrderedReady`.

Can also confirm this, but creating `StatefulSet` was not straight forward when operator created it too. Had to use a loop and delete several times:

```
while true; do oc create -f statefulset.yaml; done
```

Then `oc delete sts opensearch-masters --cascade=orphan` until log in loop showed that we created `StatefulSet`:

> Error from server (AlreadyExists): error when creating "nodes.yaml": statefulsets.apps "opensearch-nodes" already exists
> Error from server (AlreadyExists): error when creating "nodes.yaml": statefulsets.apps "opensearch-nodes" already exists
> statefulset.apps/opensearch-nodes created

**Comment by kfox1111** [comment]
> 
> Can also confirm this, but creating `StatefulSet` was not straight forward when operator created it too. Had to use a loop and delete several times:
> 
> ```
> while true; do oc create -f statefulset.yaml; done
> ```

Can scale down the operator first instead.
